### PR TITLE
Add STM32F303xB SOC description

### PR DESCRIPTION
**Pull request by mgritz**
Add a description of the STM32F303xB, based on Linaro's description of the ...303xC with smaller flash and RAM.

### CODE REVIEW

File: soc/arm/st_stm32/stm32f3/Kconfig.defconfig.stm32f303xb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# ST Microelectronics STM32F303XB MCU
+
+# Copyright (c) 2023 Martin Gritzan
+# SPDX-License-Identifier: Apache-2.0
+
+if SOC_STM32F303XB
+
+config SOC
+	default "stm32f303xb"
```
Comment (erwango) [inline soc/arm/st_stm32/stm32f3/Kconfig.defconfig.stm32f303xb:9]: This isn't obvious, but looks like it should be stm32f303xc as STM32F303xB is not an available option here:
https://github.com/zephyrproject-rtos/hal_stm32/blob/d49fb6dbe780b931eee33cce8656e87c46070fa8/stm32cube/stm32f3xx/soc/stm32f3xx.h#L60

So I rather think you should update soc/arm/st_stm32/stm32f3/Kconfig.defconfig.stm32f303xc file

File: dts/arm/st/f3/stm32f303Xb.dtsi [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	};
+
+	soc {
+		flash-controller@40022000 {
+			flash0: flash@8000000 {
+				reg = <0x08000000 DT_SIZE_K(128)>;
+			};
+		};
+
+		dma2: dma@40020400 {
```
Comment (erwango) [inline dts/arm/st/f3/stm32f303Xb.dtsi:28]: I'm a bit puzzled by duplicated information here;
I'd suggest to get dts/arm/st/f3/stm32f303Xc.dtsi inluding dts/arm/st/f3/stm32f303Xb.dtsi

**Review by erwango — CHANGES_REQUESTED** [review]

File: soc/arm/st_stm32/stm32f3/Kconfig.defconfig.stm32f303xb (hunk 1 above)
Comment (mgritz) [reply soc/arm/st_stm32/stm32f3/Kconfig.defconfig.stm32f303xb]: Good catch, here's one to the cargo cult 😅 . Now I understand how the HAL target selection comes into place.

I unified both files to a `Kconfig.defconfig.stm32f303x(b-c)`. Inclusion works by globbing and `CONFIG_SOC` is really only used for this purpose [in the HAL's CMakeLists.txt](https://github.com/zephyrproject-rtos/hal_stm32/blob/d49fb6dbe780b931eee33cce8656e87c46070fa8/stm32cube/CMakeLists.txt#L11). I find this looks better. WDYT?

File: dts/arm/st/f3/stm32f303Xb.dtsi (hunk 2 above)
Comment (mgritz) [reply dts/arm/st/f3/stm32f303Xb.dtsi:28]: Yep. I also thought about including another layer of inclusion that does all the duplicated stuff and have the ...Xb.dtsi and ...Xc.dtsi only define the RAM- and flash node, however, I did not come up with a good name for that other layer of inclusion. So as you said.

**Review by erwango — APPROVED** [review]

**Review by gautierg-st — APPROVED** [review]

**Review by FRASTM — APPROVED** [review]

**Pull request merged**